Sources/protoc-gen-swift/MessageFieldGenerator.swift

```diff
@@ -47,6 +47,10 @@ class MessageFieldGenerator: FieldGeneratorBase, FieldGenerator {
         return false
       }
     }
+  
+    var usesDefaultValueFlagForTraversal: Bool {
```

Comment:
Generally, we don't want to have to check syntax any more (working towards Editions). Would just looking at say hasFieldPresence and if the field is repeated/map vs. singular work here?

Reply:
There is a difference in behavior between proto2 and proto3 though. At least in the python lib which I was using for reference. In proto3 unset optional primitives are omitted from json, and in proto2 unset optional primitives are included. I'm not sure if theres a better way to do this.

Reply:
The `Descriptor` support should already be tracking most of that: https://github.com/apple/swift-protobuf/blob/main/Sources/SwiftProtobufPluginLibrary/Descriptor.swift#L638-L642  What's the python code you're looking at for reference?

Reply:
I didn't look at python code, just observed how python behaves when serializing a message with every kind of field type.

Reply:
Here is an example of what I mean

Proto3 behavior

```protobuf
syntax = "proto3";

message Proto3Message {
    int32 singularInt32 = 1;
    repeated int32 repeatedInt32 = 2;
    optional int32 optionalInt32 = 3;
}
```

```python
print(MessageToJson(Proto3Message(), including_default_value_fields=True))
```
```
{
  "singularInt32": 0,
  "repeatedInt32": []
}
```

Proto2 behavior
```protobuf
message Proto2Message {
    required int32 singularInt32 = 1;
    repeated int32 repeatedInt32 = 2;
    optional int32 optionalInt32 = 3;
}
```
```python
print(MessageToJson(Proto2Message(), including_default_value_fields=True))
```
```
{
  "singularInt32": 0,
  "repeatedInt32": [],
  "optionalInt32": 0
}
```

`hasPresence` will be true for `optionalInt32` for both proto2 and proto3 but we want to print it in proto2 and omit it in proto3

Reply:
I'm reaching out to the protobuf team for some advise on this to confirm the intended behavior.

Reply:
Small progress update on this - So the initial response was that they team would have expected proto2 and proto3 syntax optionals to align, but as you noticed, they don't, and there are even some tests across the languages the essentially check that they don't match.  So I'm currently trying to see what they then means, are they going to bring things in line with that they actually intended or does one need to duplicate the current behavior.

So generally, I'd say I'm pretty happen with the CL, I'd like some of the other maintainers to wait in one some of the style things I asked about also.  Realistically, given the time I hear, I fear we might not get the answer about what behavior we actually should implement until after the holiday, but I'll relay along what I find out as soon as I can.

Reply:
That sounds good, thank you for the update! I was also surprised by this behavior too honestly. IMO the _correct_ behavior for proto2 should be to include fields if they are marked as `required` and exclude unset `optional` fields.

Reply:
No update on this protobuf team side; with the holidays, that sorta makes sense. I'll relay things as soon as I hear them.

Re `required`, the name has always bugged me, because it only has meaning in the context of the _binary encoding format_.  For TextFormat (or JSON), it has no meaning, so we'll see where they end up, but those fields do have _presence_, it's just the extra validation for the binary format that really cares about `required`.

Reply:
I see okay, keep me posted on the desired functionality 🙏

Reply:
@thomasvl any updates?

Reply:
I sent another ping today asking for any updates. 🤞

Reply:
So it looks like the final direction is going to be the option should mean _Emit all without-presence fields_. So proto2 syntax `required` and `optional` don't get output, and proto3 syntax `optional` don't get output.  This would mean Python, C++, and Java aren't currently to spec, so they are scoping getting their implementations changed to match.  I've asked about when we might see a docs update to better capture the intent.

Reply:
Sounds good I can work on that. Thank you!

Reply:
Looks like a docs rollout is happening:

https://protobuf.dev/programming-guides/proto2/#json-options

> **Always emit fields without presence:** Fields that don’t support presence and that have their default value are omitted by default in JSON output (for example, an implicit presence integer with a 0 value, implicit presence string fields that are empty strings, and empty repeated and map fields). An implementation may provide an option to override this behavior and output fields with their default values.
> 
> As of v25.x, the C++, Java, and Python implementations are nonconformant, as this flag affects proto2 `optional` fields but not proto3 `optional` fields. A fix is planned for a future release.

https://protobuf.dev/programming-guides/proto3/#json-options

> **Always emit fields without presence:** Fields that don’t support presence and that have their default value are omitted by default in JSON output (for example, an implicit presence integer with a 0 value, implicit presence string fields that are empty strings, and empty repeated and map fields). An implementation may provide an option to override this behavior and output fields with their default values.

Reply:
I'm confused by those docs for proto2
> for example, an implicit presence integer with a 0 value, implicit presence string fields that are empty strings, and empty repeated and map fields

Don't integer and string fields always have presence under proto2? Under what scenario would we actually emit these in proto2?

Reply:
I've updated the logic to just use `hasPresence` for proto2 the same way as in proto3. But I think those docs could use some clarification.

Reply:
Please feel free to open an issue upstream with the feedback.  Some times it helps to hear directly from devs.